File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function(n, edges, source, destination) {
    const graph = {}
    for (let i=0; i < n; i++) {
        graph[i] = []
    }
    for (const [u, v] of edges) {
        graph[u].push(v)
        graph[v].push(u)
    }
    return dfs(source, destination, new Set())
    
    function dfs(src, dst, visiting) {
        if (src === dst) return true
        if (!visiting.has(src)) {
            visiting.add(src)
            for (const next of graph[src]) {
                if (dfs(next, dst, visiting)) return true
            }
        }
        return false
    }
};

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        
        graph = {}
        
        for i in range(n):
            graph[i] = []
        
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
            
        def dfs(src: int, dst: int, visiting: set) -> bool:
            if src == dst:
                return True
            
            if src in visiting:
                return False
            
            visiting.add(src)
            
            for next in graph[src]:
                if dfs(next, dst, visiting):
                    return True
            
            return False
        
        return dfs(source, destination, set())